### PR TITLE
Fixed horizontal padding for InputInfo and InputError

### PR DESCRIPTION
**Pull request by ZEISSLaurinAgostini**
### Prerequisites

Please make sure you can check the following two boxes:

- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project

### Contribution Type

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [x] Bug fix (non-breaking change which fixes an issue, please reference the issue id)
- [ ] New feature (non-breaking change which adds functionality, make sure to open an associated issue first)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### Description

Fixed issue https://github.com/ZEISS/precise-ui/issues/50 with correct padding for info and error

**Before**
info_old
error_old
**Now**
info_new
error_new

### CODE REVIEW

**Comment by emhaas** [comment]
For RadioButton and Checkbox the error text should be alligned with the main label text. 
Bildschirmfoto 2019-07-31 um 09 51 30
Bildschirmfoto 2019-07-31 um 09 50 28

**Comment by dimabory** [comment]
@emhaas Please take a look at the component when there is no 'label'. Should it look like this? 
https://raw.githubusercontent.com/ZEISS/precise-ui/88737477381d40de416b59210ed14b94f76f59ba/integration/__image_snapshots__/TagBuilder_9-snap.png

**Comment by ZEISSLaurinAgostini** [comment]
I would say yes. The horizontal alignment of all 3 elements (text, label, info/error) is always the same, even if just one of them is there.

**Comment by emhaas** [comment]
As this isn't often or at all the case, it can be like that!

**Comment by emhaas** [comment]
Text should be alligned with the big text but this is not a realistic use case anyway...

**Comment by dimabory** [comment]
> Text should be alligned with the big text but this is not a realistic use case anyway...

Nothing impossible and we can (should) deal with it 😉

File: src/components/Checkbox/index.tsx [hunk 1]
```diff
@@ -96,10 +77,12 @@ const RealCheckbox = styled.input`
   display: none;
 `;
 
-const FlexContainer = styled.div`
+const FlexContainer = styled.div.attrs(({ withInputInfo = false }: { withInputInfo: boolean }) => ({
```
Comment (kotielnikov) [inline src/components/Checkbox/index.tsx]: why are you using `attrs` here?

File: src/hoc/withInputInfo.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
```
Comment (kotielnikov) [inline src/hoc/withInputInfo.tsx]: You have wrong understanding of High Order Components: High order component stands for the **enhancing** of other basic component. What are you doing here is a conditional rendering, which can be easily archived with a normal component. Please refactor it to `InputNotification` and add to the components folder.

File: src/components/PaddedContainer/index.tsx [hunk 3]
```diff
@@ -32,10 +32,10 @@ interface PaddedContainerProps {
...
 const StyledContainer = styled('div')<PaddedContainerProps>(
   ({ top, left, bottom, right, gutter }) => css`
-    ${top ? `padding-top: ${distance[top]};` : ''}
-    ${left ? `padding-left: ${distance[left]};` : ''}
-    ${bottom ? `padding-bottom: ${distance[bottom]};` : ''}
-    ${right ? `padding-right: ${distance[right]};` : ''}
+    padding-top: ${(top && distance[top]) || 0};
+    padding-left: ${(left && distance[left]) || 0};
+    padding-bottom: ${(bottom && distance[bottom]) || 0};
+    padding-right: ${(right && distance[right]) || 0};
```
Comment (kotielnikov) [inline src/components/PaddedContainer/index.tsx]: why these changes are needed?

File: src/components/RadioButton/index.tsx [hunk 4]
```diff
@@ -110,9 +111,11 @@ const SelectMark = styled('div')<RadioButtonCircleProps>`
   transform: ${props => (props.selected ? 'scale(1)' : 'scale(0)')};
 `;
 
-const FlexContainer = styled.div`
+const FlexContainer = styled.div.attrs(({ withError = false }: { withError: boolean }) => ({
```
Comment (kotielnikov) [inline src/components/RadioButton/index.tsx]: why are using `attrs`?

File: src/utils/input.tsx [hunk 5]
```diff
@@ -1,21 +1,7 @@
 import * as React from 'react';
-import { InputError } from '../components/InputError';
-import { InputInfo } from '../components/InputInfo';
+import { withInputInfo } from '../hoc/withInputInfo';
 
 export function showInputInfo(error?: React.ReactChild, info?: React.ReactChild) {
```
Comment (kotielnikov) [inline src/utils/input.tsx]: export this method from the new introduced component `InputNotification`

File: src/components/Checkbox/index.tsx (hunk 1 above)
Comment (dimabory) [reply src/components/Checkbox/index.tsx]: why not?

File: src/components/PaddedContainer/index.tsx (hunk 3 above)
Comment (dimabory) [reply src/components/PaddedContainer/index.tsx]: to forcibly set padding to `0px`

File: src/hoc/withInputInfo.tsx (hunk 2 above)
Comment (dimabory) [reply src/hoc/withInputInfo.tsx]: `InputNotification` cannot be a react component. 
Just take a look what's returning by `withInputInfo` (and I think it's kinda enhancing).

File: src/components/Checkbox/index.tsx (hunk 1 above)
Comment (kotielnikov) [reply src/components/Checkbox/index.tsx]: Please read the documentation 
https://www.styled-components.com/docs/faqs#when-to-use-attrs

File: src/hoc/withInputInfo.tsx (hunk 2 above)
Comment (kotielnikov) [reply src/hoc/withInputInfo.tsx]: Please read carefully the documentation:
https://reactjs.org/docs/higher-order-components.html

> Concretely, a higher-order component is a function that takes a component and returns a new component.

It's called **enhancing** .

File: src/components/PaddedContainer/index.tsx (hunk 3 above)
Comment (kotielnikov) [reply src/components/PaddedContainer/index.tsx]: Thanks, I know what `padding` css property means. The question was why do you need to set it forcibly?

File: src/hoc/withInputInfo.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      </InputError>
+    ) : (
+      <>{error}</>
+    );
+  };
+
+  if (error) return (props: InputErrorProps) => <WithError error={error} {...props} />;
+  if (info) return (props: InputInfoProps) => <WithInfo info={info} {...props} />;
+
+  return () => <></>;
```
Comment (kotielnikov) [inline src/hoc/withInputInfo.tsx]: React.Fragments are for different purpose - do group several react nodes. If you don't have nodes, then just not use fragment. use **null**

File: src/hoc/withInputInfo.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    return typeof error === 'string' ? (
+      <InputError left="medium" {...props}>
+        {error}
+      </InputError>
+    ) : (
+      <>{error}</>
+    );
+  };
+
+  if (error) return (props: InputErrorProps) => <WithError error={error} {...props} />;
```
Comment (kotielnikov) [inline src/hoc/withInputInfo.tsx]: if/else one-liners are agains our style-guides

File: src/hoc/withInputInfo.tsx (hunk 2 above)
Comment (dimabory) [reply src/hoc/withInputInfo.tsx]: Can we consider `React.ReactChild` as a component?

File: src/components/PaddedContainer/index.tsx (hunk 3 above)
Comment (dimabory) [reply src/components/PaddedContainer/index.tsx]: To override default props if it is existing. For instance, take a look at https://github.com/ZEISS/precise-ui/pull/139/files?file-filters%5B%5D=.md&file-filters%5B%5D=.png&file-filters%5B%5D=.tsx#diff-98434080865c9ea31f591a968fbe9909

By default `InputInfo` has `padding-left="medium"` and now it is possible to  forcibly reset this value.

File: src/hoc/withInputInfo.tsx (hunk 7 above)
Comment (dimabory) [reply src/hoc/withInputInfo.tsx]: can you provide me the link to the style-guide?

File: src/hoc/withInputInfo.tsx (hunk 2 above)
Comment (kotielnikov) [reply src/hoc/withInputInfo.tsx]: React component:
```
const Component = () => <div />
```

ReactChild, which is also a `ReactElement | string | number`:
```
<Component />
```

A good article describing principles:
https://medium.com/@fay_jai/react-elements-vs-react-components-vs-component-backing-instances-14d42729f62

Comment (kotielnikov) [reply src/hoc/withInputInfo.tsx]:
```

export interface InputNotificationProps extends InputErrorProps, InputInfoProps {
  error?: React.ReactNode;
  info?: React.ReactNode;
}

export const InputNotification: React.FC<InputNotificationProps> = ({ error, info, ...rest }) => {
  if (error) {
    if (typeof error === 'string') {
      return <InputError {...rest}>{error}</InputError>;
    }

    return <>{error}</>;
  } else if (info) {
    if (typeof info === 'string') {
      return <InputInfo {...rest}>{info}</InputInfo>;
    }

    return <>{info}</>;
  }

  // tslint:disable-next-line
  return null;
};

/**
 * @deprecated use InputNotification component instead
 * @param error ReactChild
 * @param info ReactChild
 */
export function showInputInfo(error?: React.ReactChild, info?: React.ReactChild) {
  return <InputNotification error={error} info={info} />;
}
```

File: src/components/InputError/index.tsx [hunk 8]
```diff
@@ -1,21 +1,13 @@
...
-  /**
-   * The content of input error (error to display).
-   */
-  children?: React.ReactNode;
-}
+export interface InputErrorProps extends PaddedContainerProps {}
 
-const StyledError = styled('div')<InputErrorProps>`
-  ${getFontStyle({ size: 'xSmall' })}
+const StyledError = styled(PaddedContainer)<InputErrorProps>`
```
Comment (kotielnikov) [inline src/components/InputError/index.tsx]: `styled.div`

I think there was a misunderstanding. Gutters should be up to the consumer. Semantically error/info notification style has nothing to do with `PaddedContainer`, while  `PaddedContainer` is a container and Info/Error Input is styling of error notifications. they serve different purposes. It has no sense that this component has the container props.

When you need to add gutters to InputNotification you will use composition either through styled-components api or just as a basic composition

i.e
```
// through styled-components
const NotificationWithGutter = styled(InputNotification)`
    padding: ${distance.small};
`
```
or
```
// standard composition
<PaddedContainer><InputNotification {....} /></PaddedContainer>
```

File: src/components/InputInfo/index.tsx [hunk 9]
```diff
@@ -1,20 +1,13 @@
...
-  /**
-   * The content of input info (text to display).
-   */
-  children?: React.ReactNode;
-}
+export interface InputInfoProps extends PaddedContainerProps {}
 
-const StyledInfo = styled('div')<InputInfoProps>`
-  ${getFontStyle({ size: 'xSmall' })}
+const StyledInfo = styled(PaddedContainer)<InputInfoProps>`
```
Comment (kotielnikov) [inline src/components/InputInfo/index.tsx]: styled.div
